### ISSUE TITLE
Doubly Parameterized tests within a nested test class

### ISSUE THREAD

**Issue by xenoterracide** [junit-team/junit5#1456]
I'm not sure if this is a doc improvement or request for feature, from the docs I'm not sure if what I'm trying to do is possible which is why I'm posting here

```java
class XenoPluginTest {

    private Project project;

    @BeforeEach
    void setup() {
        project = ProjectBuilder.builder().build();
    }

    @ParameterizedTest
    @ArgumentsSource( PluginProvider.class )
    void codeQuality( Collection<Class<Plugin<Project>>> pluginClass ) {
        pluginClass.forEach( project.getPluginManager()::apply );

        PluginContainer plugins = project.getPlugins();

        assertAll( "plugins",
                () -> assertTrue( plugins.hasPlugin( IdeaPlugin.class ), "idea" ),
                () -> assertTrue( plugins.hasPlugin( SpotBugsPlugin.class ), "spotbugs" ),
                () -> assertTrue( plugins.hasPlugin( ErrorPronePlugin.class ), "errorprone" ),
                () -> assertTrue( plugins.hasPlugin( CheckstylePlugin.class ), "checkstyle" ),
                () -> assertTrue( plugins.hasPlugin( JavaLibraryPlugin.class ), "java" )
        );
    }

    @Nested
    class SourceSetExtensions {

        @ParameterizedTest
        @ArgumentsSource( ExtensionProvider.class )
        void hasSourceSets( Class<? extends CodeQualityExtension> cqe ) {
            assertThat( project.getExtensions().getByType( cqe ).getSourceSets() ).isNotEmpty();
        }
    }

    static class ExtensionProvider implements ArgumentsProvider {

        @Override
        public Stream<? extends Arguments> provideArguments( ExtensionContext context ) {
            return Stream.of( SpotBugsExtension.class, CheckstyleExtension.class ).map( Arguments::of );
        }
    }

    static class PluginProvider implements ArgumentsProvider {

        @Override
        public Stream<? extends Arguments> provideArguments( ExtensionContext context ) {
            return Stream.of(
                    Arrays.asList( JavaLibraryPlugin.class, XenoCodeQualityPlugin.class ),
                    Collections.singleton( XenoJavaLibPlugin.class )
            ).map( Arguments::of );
        }
    }
}
```

What I really want here, `codeQuality` runs twice, once for each plugin class list (this works). `hasSourceSets` runs *4* times. The `@Nested` runs twice for each plugin class, and then `hasSourceSets` would run twice fore each of those with each extension given. If this is currently possible perhaps an example could be added to the docs. If it isn't possible, I think being able to do this somehow would be a great feature.

**Comment by xenoterracide** [comment]
I think so? at the very least related. I believe the solution proposed in #871  would work, that would make my test look like this?

```
@ArgumentsSource( PluginProvider.class )
class XenoPluginTest {

    private Project project;

    @BeforeEach
    // @ArgumentsSource( PluginProvider.class ) hmm does this work now?
    void setup( Collection<Class<Plugin<Project>>> pluginClass ) {
        project = ProjectBuilder.builder().build();
        pluginClass.forEach( project.getPluginManager()::apply );
    }

    @ParameterizedTest
    void codeQuality(  ) {

        PluginContainer plugins = project.getPlugins();

        assertAll( "plugins",
                () -> assertTrue( plugins.hasPlugin( IdeaPlugin.class ), "idea" ),
                () -> assertTrue( plugins.hasPlugin( SpotBugsPlugin.class ), "spotbugs" ),
                () -> assertTrue( plugins.hasPlugin( ErrorPronePlugin.class ), "errorprone" ),
                () -> assertTrue( plugins.hasPlugin( CheckstylePlugin.class ), "checkstyle" ),
                () -> assertTrue( plugins.hasPlugin( JavaLibraryPlugin.class ), "java" )
        );
    }

    @Nested
    class SourceSetExtensions {

        @ParameterizedTest
        @ArgumentsSource( ExtensionProvider.class )
        void hasSourceSets( Class<? extends CodeQualityExtension> cqe ) {
            assertThat( project.getExtensions().getByType( cqe ).getSourceSets() ).isNotEmpty();
        }
    }

    static class ExtensionProvider implements ArgumentsProvider {

        @Override
        public Stream<? extends Arguments> provideArguments( ExtensionContext context ) {
            return Stream.of( SpotBugsExtension.class, CheckstyleExtension.class ).map( Arguments::of );
        }
    }

    static class PluginProvider implements ArgumentsProvider {

        @Override
        public Stream<? extends Arguments> provideArguments( ExtensionContext context ) {
            return Stream.of(
                    Arrays.asList( JavaLibraryPlugin.class, XenoCodeQualityPlugin.class ),
                    Collections.singleton( XenoJavaLibPlugin.class )
            ).map( Arguments::of );
        }
    }
}
```

**Comment by sbrannen** [comment]
Actually, I don't think it would look like that.

Take a look at the proposal in https://github.com/junit-team/junit5/issues/878#issuecomment-354544841.

What I was alluding to is that you parameterize the container/class and access the "arguments" via `@Param` (or similar) via fields. Then you'd have a `@ParameterizedTest` method directly in that class that would do what you are currently proposing to do in a `@Nested` test class.

In other words, I think the combination of a parameterized container and parameterized test would obviate the need for a nested test class.

**Comment by sbrannen** [comment]
By the way, your `ExtensionProvider` is overkill.

You can achieve that simply via the following:

```java
@ValueSource( classes = { SpotBugsExtension.class, CheckstyleExtension.class } )
```

**Comment by marcphilipp** [comment]
Duplicate of #878
